pulsar-broker/src/main/java/org/apache/pulsar/compaction/StrategicTwoPhaseCompactor.java

```diff
@@ -418,10 +418,9 @@ private <T> void phaseTwoLoop(String topic, Iterator<Message<T>> reader,
                                 .whenComplete((res, exception2) -> {
                                     if (exception2 != null) {
                                         promise.completeExceptionally(exception2);
-                                        return;
                                     }
+                                    phaseTwoLoop(topic, reader, lh, outstanding, promise);
                                 });
-                        phaseTwoLoop(topic, reader, lh, outstanding, promise);
```

Comment:
The `RawBatchMessageContainerImpl ` is not thread-safe, right? If we add entries parallel, I wonder if this will cause data loss?(add message and clear at same time)

Reply:
Probably, we should use the raw reader to do this, and we have to compact each batch into different entries.

Reply:
> Probably we should use the raw reader to do this, and we have to compact each batch into different entries.

We could do, but we need to get values from the RawReader to apply the compaction strategy. We are using the reader interface to reuse the value-parsing logic.